### PR TITLE
netavark, dns: don't `double-fork` aardvark instead wait for the aardvark process to return.

### PR DESCRIPTION
**Pull request by flouthoc**
Netavark now does not double-forks aardvark-dns's server instead it
waits for aardvark-process to return back and success return means
aardvark-dns is ready to serve requests and now forking happens at
aardvark end.

See: https://doc.rust-lang.org/std/process/struct.Command.html#method.spawn

This needs: https://github.com/containers/aardvark-dns/pull/148

Should help in:
* https://github.com/containers/podman/issues/14173
* https://github.com/containers/podman/issues/14171

Alternative to: https://github.com/containers/netavark/pull/300

### CODE REVIEW

**Comment by flouthoc** [comment]
@Luap99 I tried this in combination with https://github.com/containers/aardvark-dns/pull/148 and setup works fine for me. Now aardvark will return from parent when everything is ready and it is actually serving DNS requests on the assigned IP addresses ( only tries for 10 times with delay of 10ms ).

File: src/dns/aardvark.rs [hunk 1]
```diff
@@ -106,14 +106,17 @@ impl Aardvark {
             false => Stdio::null(),
         };
 
+        // After https://github.com/containers/aardvark-dns/pull/148 this command
+        // will block till aardvark-dns's parent process returns back.
         Command::new(&aardvark_args[0])
             .args(&aardvark_args[1..])
             .stdin(Stdio::null())
             .stdout(Stdio::null())
             .stderr(output)
```
Comment (Luap99) [inline src/dns/aardvark.rs:115]: We should pass the fds to aardvark and let aardvark close(dup them to /dev/null) when it is ready

Comment (flouthoc) [reply src/dns/aardvark.rs:115]: Sure we could do that.

Comment (flouthoc) [reply src/dns/aardvark.rs]: Added in latest push and tested both the PR with podman

File: src/dns/aardvark.rs [hunk 2]
```diff
@@ -101,19 +100,18 @@ impl Aardvark {
...
-            .stdout(Stdio::null())
-            .stderr(output)
+            .stdin(Stdio::inherit())
+            .stdout(Stdio::inherit())
+            .stderr(Stdio::inherit())
             // set RUST_LOG for aardvark
             .env("RUST_LOG", log::max_level().as_str())
-            .spawn()?;
+            .output()
+            .expect("failed to start aardvark-dns process");
```
Comment (Luap99) [inline src/dns/aardvark.rs:114]: do not panic! please return a proper error

Comment (flouthoc) [reply src/dns/aardvark.rs:114]: I think `?` will return error from the function. Did you mean wrapping the error.

Comment (flouthoc) [reply src/dns/aardvark.rs]: fixed.

**Comment by flouthoc** [comment]
Rebased.

**Comment by Luap99** [comment]
LGTM I think it is expected that the integration tests fail since we need the aardvark from the other PR.

**Comment by baude** [comment]
@flouthoc i'll take a peek

**Comment by flouthoc** [comment]
( Just pasting here for info ) To make sure i ran tests on my local ( but my local could be different than CI) . So maybe we should confirm if `aardvark` is updated on CI or not.
```console
[fl@fedora netavark]$ sudo bats test/
[sudo] password for fl: 
 ✓ netavark version
 ✓ netavark error - invalid ns path
 ✓ netavark error - invalid config path
 ✓ check iptables driver is in use
 ✓ iptables - internal network
 ✓ iptables - simple bridge
 ✓ iptables - ipv6 bridge
 ✓ iptables - dual stack dns with alt port
 ✓ iptables - check error message from netns thread
 ✓ iptables - port forwarding ipv4 - tcp
 ✓ iptables - port forwarding ipv6 - tcp
 ✓ iptables - port forwarding dualstack - tcp
 ✓ iptables - port forwarding ipv4 - udp
 ✓ iptables - port forwarding ipv6 - udp
 ✓ iptables - port forwarding dualstack - udp
 ✓ iptables - port forwarding ipv4 - sctp
 ✓ iptables - port forwarding ipv6 - sctp
 ✓ iptables - port forwarding dualstack - sctp
 ✓ iptables - port range forwarding ipv4 - tcp
 ✓ iptables - port range forwarding ipv6 - tcp
 ✓ iptables - port range forwarding ipv4 - udp
 ✓ iptables - port range forwarding ipv6 - udp
 ✓ iptables - port range forwarding dual - udp
 ✓ iptables - port range forwarding dual - tcp
 ✓ iptables - port forwarding with hostip ipv4 - tcp
 ✓ iptables - port forwarding with hostip ipv4 dual stack- tcp
 ✓ iptables - port range forwarding with hostip ipv6 - tcp
 ✓ iptables - port range forwarding with hostip ipv6 dual stack - tcp
 ✓ iptables - port range forwarding with hostip ipv4 - udp
 ✓ iptables - port range forwarding with hostip ipv6 - udp
 ✓ bridge ipam none
 ✓ bridge unknown ipam driver
 ✓ iptables - isolate networks
 ✓ iptables - test read only /proc
 - check firewalld driver is in use (skipped: TODO: Firewalld driver swapped with iptables until firewalld 1.1.0)
 ✓ firewalld - simple bridge
 ✓ firewalld - ipv6 bridge
 ✓ firewalld - dual stack dns with alt port
 ✓ firewalld - check error message from netns thread
 ✓ firewalld - port forwarding ipv4 - tcp
 ✓ firewalld - port forwarding ipv6 - tcp
 ✓ firewalld - port forwarding dualstack - tcp
 ✓ firewalld - port forwarding ipv4 - udp
 /usr/libexec/bats-core/bats-exec-test: line 116: 39038 Killed                  nsenter -n -t $HOST_NS_PID firewalld --nopid --nofork --system-config "$NETAVARK_TMPDIR" &> "$NETAVARK_TMPDIR/firewalld.log"                            44/65
 ✓ firewalld - port forwarding ipv6 - udp
 ✓ firewalld - port forwarding dualstack - udp
 /usr/libexec/bats-core/bats-exec-test: line 161: 39742 Killed                  nsenter -n -t $HOST_NS_PID firewalld --nopid --nofork --system-config "$NETAVARK_TMPDIR" &> "$NETAVARK_TMPDIR/firewalld.log"                            46/65
 ✓ firewalld - port forwarding ipv4 - sctp
 ✓ firewalld - port forwarding ipv6 - sctp
 /usr/libexec/bats-core/bats-exec-test: line 159: 40353 Killed                  nsenter -n -t $HOST_NS_PID firewalld --nopid --nofork --system-config "$NETAVARK_TMPDIR" &> "$NETAVARK_TMPDIR/firewalld.log"                            48/65
 ✓ firewalld - port forwarding dualstack - sctp
 ✓ firewalld - port range forwarding ipv4 - tcp
 /usr/libexec/bats-core/bats-exec-test: line 173: 41079 Killed                  nsenter -n -t $HOST_NS_PID firewalld --nopid --nofork --system-config "$NETAVARK_TMPDIR" &> "$NETAVARK_TMPDIR/firewalld.log"                            50/65
 ✓ firewalld - port range forwarding ipv6 - tcp
 /usr/libexec/bats-core/bats-exec-test: line 102: 41406 Killed                  nsenter -n -t $HOST_NS_PID firewalld --nopid --nofork --system-config "$NETAVARK_TMPDIR" &> "$NETAVARK_TMPDIR/firewalld.log"                            51/65
 ✓ firewalld - port range forwarding ipv4 - udp
 ✓ firewalld - port range forwarding ipv6 - udp
 ✓ firewalld - port range forwarding dual - udp
 ✓ firewalld - port range forwarding dual - tcp
 ✓ firewalld - port forwarding with hostip ipv4 - tcp
 /usr/libexec/bats-core/bats-exec-test: line 161: 43286 Killed                  nsenter -n -t $HOST_NS_PID firewalld --nopid --nofork --system-config "$NETAVARK_TMPDIR" &> "$NETAVARK_TMPDIR/firewalld.log"                            56/65
 ✓ firewalld - port forwarding with hostip ipv4 dual stack- tcp
 ✓ firewalld - port range forwarding with hostip ipv6 - tcp
 ✓ firewalld - port range forwarding with hostip ipv6 dual stack - tcp
 /usr/libexec/bats-core/bats-exec-test: line 114: 44367 Killed                  nsenter -n -t $HOST_NS_PID firewalld --nopid --nofork --system-config "$NETAVARK_TMPDIR" &> "$NETAVARK_TMPDIR/firewalld.log"                            59/65
 ✓ firewalld - port range forwarding with hostip ipv4 - udp
 ✓ firewalld - port range forwarding with hostip ipv6 - udp
 ✓ simple macvlan setup
 ✓ macvlan setup internal
 ✓ macvlan setup with mtu
 ✓ macvlan modes
 ✓ macvlan ipam none

65 tests, 0 failures, 1 skipped
```

**Comment by cevich** [comment]
> I am sure that aardvark is not updated, you could check with hack/get_ci_vm

Yep.  So it's suppose to come from CI runs on the 'main' branch, perhaps that is broken.  [You can see by going to `https://cirrus-ci.com/github/containers/aardvark-dns/<branch>`](https://cirrus-ci.com/github/containers/aardvark-dns/main)...Looks mostly green :confused: 

We do collect an `aardvark-dns.info` file in the zip, that should contain details about the build where the binary came from...

**Comment by baude** [comment]
already fixed in #344

**Comment by cevich** [comment]
Yep @baude we need that.  I'll work on fixing the images as well (for the future).

**Comment by flouthoc** [comment]
aha, integration tests are green with new aardvark. @Luap99 @baude @mheon This should be good to go now.

**Pull request merged**